### ISSUE TITLE
Understanding Hinton’s Capsule Networks. Part IV: CapsNet Architecture

### ISSUE THREAD

**Issue by sweaterr** [sweaterr/readinglist#343]
https://medium.com/@pechyonkin/part-iv-capsnet-architecture-6a64422f7dce

**Comment by sweaterr** [comment]
## Introduction

In this part, I will walk through the architecture of the CapsNet.
이 부분에서는 CapsNet의 아키텍처를 살펴 보겠습니다.

I will also offer my shot at calculating the number of trainable parameters in the CapsNet.
나는 또한 CapsNet에서 훈련 가능한 매개 변수의 수를 계산할 때 나의 기회를 제공 할 것이다.

My resulting number is around 8.2 million of trainable parameters which is different from the 11.36 officially referred to in the paper.
결과적으로 약 820 만 개의 훈련 가능한 매개 변수가 공식적으로이 11.36과는 다릅니다.

The paper itself is not very detailed and hence it leaves some open questions about specifics of the network implementation that are as of today still unanswered because the authors did not provide their code.
논문 그 자체는 그리 상세하지 않기 때문에 저자가 자신의 코드를 제공하지 않았기 때문에 오늘날까지도 아직 답변이없는 네트워크 구현의 특성에 대한 공개적인 질문을 남깁니다.

Nonetheless, I still think that counting parameters in a network is a good exercise for purely learning purposes as it allows one to practice understanding of all building blocks of a particular architecture.
그럼에도 불구하고 네트워크의 매개 변수를 세는 것은 특정 아키텍처의 모든 빌딩 블록에 대한 이해를 연습 할 수 있기 때문에 순전히 학습 목적으로 좋은 연습이라고 생각합니다.

The CapsNet has 2 parts: encoder and decoder.
CapsNet에는 인코더와 디코더의 두 부분이 있습니다.

The first 3 layers are encoder, and the second 3 are decoder:  Layer 1.
처음 3 개의 레이어는 인코더이고, 2 번째 레이어는 디코더 : 레이어 1입니다.

Convolutional layer Layer 2.
컨벌루션 레이어 레이어 2.

PrimaryCaps layer Layer 3.
PrimaryCaps 레이어 레이어 3.

DigitCaps layer Layer 4.
DigitCaps 레이어 레이어 4.

Fully connected #1 Layer 5.
완전히 연결된 # 1 Layer 5.

Fully connected #2 Layer 6.
완전히 연결된 # 2 레이어 6.

Fully connected #3.
완전히 연결된 # 3.

**Comment by sweaterr** [comment]
## Part I. Encoder.

![image](https://user-images.githubusercontent.com/1518919/38616787-2867f97e-3dd0-11e8-91ed-8611d7116ea3.png)


Encoder part of the network takes as input a 28 by 28 MNIST digit image and learns to encode it into a 16-dimensional vector of instantiation parameters (as explained in the previous posts of this series), this is where the capsules do their job.
네트워크의 인코더 부분은 28 x 28 MNIST 디지트 이미지를 입력으로 받아 들여 인스턴스화 매개 변수의 16 차원 벡터로 인코딩하는 것을 배우며 (이 시리즈의 이전 게시물에서 설명) 캡슐이 작업을 수행합니다.

The output of the network during prediction is a 10-dimensional vectors of lengths of DigitCaps’ outputs.
예측 중에 네트워크의 출력은 DigitCaps의 출력 길이의 10 차원 벡터입니다.

The decoder has 3 layers: two of them are convolutional and the last one is fully connected.
디코더에는 3 개의 레이어가 있습니다. 그 중 2 개는 길쌈이고 마지막 하나는 완전히 연결됩니다.

**Comment by sweaterr** [comment]
## Layer 1. Convolutional layer

Input: 28x28 image (one color channel).
Output: 20x20x256 tensor.
Number of parameters: 20992.


Convolutional layer’s job is to detect basic features in the 2D image.
컨볼 루션 레이어의 작업은 2D 이미지의 기본 기능을 감지하는 것입니다.

In the CapsNet, the convolutional layer has 256 kernels with size of 9x9x1 and stride 1, followed by ReLU activation.
CapsNet에서 컨볼 루션 계층은 9x9x1 크기와 스트라이드 1 크기를 가진 256 개의 커널을 가지고 있으며, ReLU 활성화가 뒤 따른다.

If you don’t know what this means, here are some awesome resources that will allow you to quickly pick up key ideas behind convolutions.
이것이 무엇을 의미하는지 모른다면, 여기에 회귀 뒤의 핵심 아이디어를 빠르게 선택할 수있는 멋진 리소스가 있습니다.

To calculate the number of parameters, we need to also remember that each kernel in a convolutional layer has 1 bias term.
매개 변수의 수를 계산하려면 컨볼 루션 계층의 각 커널에 1 개의 바이어스 항이 있음을 기억해야합니다.

Hence this layer has (9x9 + 1)x256 = 20992 trainable parameters in total,.
따라서이 계층은 합계로 (9x9 + 1) x256 = 20992 개의 조정 가능한 매개 변수를 갖습니다.

**Comment by sweaterr** [comment]
## Layer 2. PrimaryCaps layer

Input: 20x20x256 tensor.
Output: 6x6x8x32 tensor.
Number of parameters: 5308672.

This layer has 32 primary capsules whose job is to take basic features detected by the convolutional layer and produce combinations of the features.
이 레이어에는 32 개의 기본 캡슐이 있으며 컨볼 루션 레이어에서 감지 한 기본 기능을 사용하여 기능 조합을 생성합니다.

The layer has 32 “primary capsules” that are very similar to convolutional layer in their nature.
이 레이어는 본질적으로 길쌈 (convolutional) 레이어와 매우 유사한 32 개의 "기본 캡슐"을 가지고 있습니다.

Each capsule applies eight 9x9x256 convolutional kernels to the 20x20x256 input volume and therefore produces 6x6x8 output tensor.
각 캡슐은 20x20x256 입력 볼륨에 8 개의 9x9x256 컨볼 루션 커널을 적용하므로 6x6x8 출력 텐서를 생성합니다.

Since there are 32 such capsules, the output volume has shape of 6x6x8x32.
이러한 캡슐이 32 개 있기 때문에 출력 볼륨의 모양은 6x6x8x32입니다.

Doing calculation similar to the one in the previous layer, we get 5308672 trainable parameters in this layer.
이전 계층의 계산과 비슷한 계산을 수행하면이 계층에 5308672 개 매개 변수가 생깁니다.

**Comment by sweaterr** [comment]
## Layer 3. DigitCaps layer

Input: 6x6x8x32 tensor.
Output: 16x10 matrix.
Number of parameters: 1497600.

This layer has 10 digit capsules, one for each digit.
이 레이어에는 각 자릿수에 대해 10 자리 캡슐이 있습니다.

Each capsule takes as input a 6x6x8x32 tensor.
각 캡슐은 입력으로 6x6x8x32 텐서를 사용합니다.

You can think of it as 6x6x32 8-dimensional vectors, which is 1152 input vectors in total, As per the inner workings of the capsule (as described here), each of these input vectors gets their own 8x16 weight matrix that maps 8-dimensional input space to the 16-dimensional capsule output space.
이것을 총 1152 개의 입력 벡터 인 6x6x32 8 차원 벡터로 생각할 수 있습니다. 여기에 설명 된 바와 같이 캡슐의 내부 동작에 따라 각 입력 벡터는 8 차원 16 진수 행렬을 가져옵니다. 입력 공간을 16 차원 캡슐 출력 공간에 연결한다.

So, there are 1152 matrices for each capsule, and also 1152 c coefficients and 1152 b coefficients used in the dynamic routing.
따라서 각 캡슐에는 1152 개의 행렬이 있고 동적 라우팅에 사용되는 1152 개의 계수와 1152 개의 계수가 있습니다.

Multiplying: 1152 x 8 x 16 + 1152 + 1152, we get 149760 trainable parameters per capsule, then we multiply by 10 to get the final number of parameters for this layer.
곱하기 : 1152 x 8 x 16 + 1152 + 1152, 캡슐 당 149760 개의 학습 가능한 매개 변수를 얻은 다음이 레이어의 최종 매개 변수 수를 얻기 위해 10을 곱합니다.

**Comment by sweaterr** [comment]
## The loss function

The loss function might look complicated at first sight, but it really is not.
손실 기능은 첫눈에 복잡해 보일 수도 있지만 실제로는 그렇지 않습니다.

It is very similar to the SVM loss function.
이는 SVM 손실 함수와 매우 유사합니다.

In order to understand the main idea about how it works, recall that the output of the DigitCaps layer is 10 sixteen-dimensional vectors.
어떻게 작동하는지에 대한 주요 아이디어를 이해하기 위해 DigitCaps 레이어의 출력은 16 개의 16 진법 벡터임을 상기하십시오.

During training, for each training example, one loss value will be calculated for each of the 10 vectors according to the formula below and then the 10 values will be added together to calculate the final loss.
트레이닝 중에 각 트레이닝 예제에 대해 아래 수식에 따라 10 개의 벡터 각각에 대해 하나의 손실 값을 계산 한 다음 10 개의 값을 더하여 최종 손실을 계산합니다.

Because we are talking about supervised learning, each training example will have the correct label, in this case it will be a ten-dimensional one-hot encoded vector with 9 zeros and 1 one at the correct position.
우리가 감독 학습에 대해 이야기하고 있기 때문에, 각 훈련 예제는 올바른 레이블을 가지게됩니다.이 경우, 정확한 위치에 9 개의 0과 1이있는 10 차원 1 핫 코드 벡터가됩니다.

In the loss function formula, the correct label determines the value of T_c: it is 1 if the correct label corresponds with the digit of this particular DigitCap and 0 otherwise.
손실 함수 식에서 올바른 레이블은 T_c의 값을 결정합니다. 올바른 레이블이이 특정 DigitCap의 숫자와 일치하면 1이고 그렇지 않으면 0입니다.

**Comment by sweaterr** [comment]
## Part II. Decoder.
![image](https://user-images.githubusercontent.com/1518919/38619433-1a82f564-3dd7-11e8-8fd9-761fb6e8211e.png)

**Comment by sweaterr** [comment]
Decoder takes a 16-dimensional vector from the correct DigitCap and learns to decode it into an image of a digit (note that it only uses the correct DigitCap vector during training and ignores the incorrect ones).
Decoder는 올바른 DigitCap에서 16 차원 벡터를 가져 와서 숫자의 이미지로 디코드하는 것을 배웁니다. (정확한 DigitCap 벡터 만 사용하고 잘못된 것을 무시합니다.)

Decoder is used as a regularizer, it takes the output of the correct DigitCap as input and learns to recreate an 28 by 28 pixels image, with the loss function being Euclidean distance between the reconstructed image and the input image.
디코더는 정규화 장치로 사용되며 올바른 DigitCap의 출력을 입력으로 사용하여 28x28 픽셀 이미지를 다시 생성하는 방법을 학습합니다. 손실 함수는 재구성 된 이미지와 입력 이미지 사이의 유클리드 거리입니다.

Decoder forces capsules to learn features that are useful for reconstructing the original image.
디코더는 캡슐이 원본 이미지를 재구성하는 데 유용한 기능을 학습하도록합니다.

The closer the reconstructed image to the input image, the better.
재구성 된 이미지가 입력 이미지에 가까울수록 좋습니다.

Examples of reconstructed images can be seen in the image below.
재구성 된 이미지의 예는 아래 이미지에서 볼 수 있습니다.

**Comment by sweaterr** [comment]
![image](https://user-images.githubusercontent.com/1518919/38619610-7f363be2-3dd7-11e8-8b22-4fcae3a8f98d.png)

**Comment by sweaterr** [comment]
## Layer 4. Fully connected #1

Input: 16x10.
Output: 512.
Number of parameters: 82432.

Each output of the lower level gets weighted and directed into each neuron of the fully connected layer as input. Each neuron also has a bias term. For this layer there are 16x10 inputs that are all directed to each of the 512 neurons of this layer. Therefore, there are (16x10 + 1)x512 trainable parameters.

For the following two layers calculation is the same: number of parameters = (number of inputs + bias) x number of neurons in the layer. This is why there is no explanation for fully connected layers 2 and 3.

**Comment by sweaterr** [comment]
Layer 5. Fully connected #2
Input: 512.
Output: 1024.
Number of parameters: 525312.

Layer 6. Fully connected #3
Input: 1024.
Output: 784 (which after reshaping gives back a 28x28 decoded image).
Number of parameters: 803600.

Total number of parameters in the network: 8238608.

**Comment by sweaterr** [comment]
## Conclusion

This wraps up the series on the CapsNet.
이것은 CapsNet에서 시리즈를 마무리합니다.

There are many very good resources around the internet.
인터넷에는 아주 좋은 자원이 많이 있습니다.

If you would like to learn more on this fascinating topic, please have a look at this awesome compilation of links about CapsNets.
이 매혹적인 주제에 대해 더 자세히 알고 싶다면 CapsNets에 대한 링크의 멋진 편집을 살펴보십시오
